### PR TITLE
RevokedCertificateBuilder extensions

### LINKED ISSUE #2533: Add feature to create CRLs
I'd like to be able to create CRLs.

`<reaperhulk> Wulf: that's a missing feature right now. A CertificateRevocationListBuilder needs to be implemented`

### PR DESCRIPTION
**Pull request by reaperhulk**
refs and fixes #2533

### CODE REVIEW

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 1]
```diff
@@ -164,6 +164,68 @@ def _encode_crl_number(backend, crl_number):
...
+    x509.ReasonFlags.key_compromise: 1,
+    x509.ReasonFlags.ca_compromise: 2,
+    x509.ReasonFlags.affiliation_changed: 3,
+    x509.ReasonFlags.superseded: 4,
+    x509.ReasonFlags.cessation_of_operation: 5,
+    x509.ReasonFlags.certificate_hold: 6,
+    x509.ReasonFlags.remove_from_crl: 8,
+    x509.ReasonFlags.privilege_withdrawn: 9,
+    x509.ReasonFlags.aa_compromise: 10
+}
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: I'm quite certain this dict already exists somewhere?

Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: There's the opposite (int to enum) in x509.py, and there's _CRLREASONFLAGS in backend.py, but the integer mappings are different (even though the names are the same) between how they're encoded in CRLDistributionPoints vs CRLReason inside a CRL entry.

Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Can we at least put this text to to the int-to-enum one?

Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: For reference:

```

   CRLReason ::= ENUMERATED {
        unspecified             (0),
        keyCompromise           (1),
        cACompromise            (2),
        affiliationChanged      (3),
        superseded              (4),
        cessationOfOperation    (5),
        certificateHold         (6),
             -- value 7 is not used
        removeFromCRL           (8),
        privilegeWithdrawn      (9),
        aACompromise           (10) }
```

```
ReasonFlags ::= BIT STRING {
     unused                  (0),
     keyCompromise           (1),
     cACompromise            (2),
     affiliationChanged      (3),
     superseded              (4),
     cessationOfOperation    (5),
     certificateHold         (6),
     privilegeWithdrawn      (7),
     aACompromise            (8) }
```

We built a ReasonFlags enum that contains the union of those two, but then we map to the correct underlying bit value using these dicts.

Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: next*

File: src/cryptography/hazmat/backends/openssl/x509.py [hunk 2]
```diff
@@ -694,13 +706,27 @@ def _decode_inhibit_any_policy(backend, asn1_int):
...
+    x509.ReasonFlags.key_compromise: 1,
+    x509.ReasonFlags.ca_compromise: 2,
+    x509.ReasonFlags.affiliation_changed: 3,
+    x509.ReasonFlags.superseded: 4,
+    x509.ReasonFlags.cessation_of_operation: 5,
+    x509.ReasonFlags.certificate_hold: 6,
+    x509.ReasonFlags.remove_from_crl: 8,
+    x509.ReasonFlags.privilege_withdrawn: 9,
+    x509.ReasonFlags.aa_compromise: 10
+}
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/x509.py]: Can we split just adding this dict into its own PR?

**Comment by alex** [comment]
Ready for the next rebase :-)

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 3]
```diff
@@ -164,6 +165,54 @@ def _encode_crl_number(backend, crl_number):
...
+    pp = backend._ffi.new('unsigned char **')
+    r = backend._lib.i2d_ASN1_GENERALIZEDTIME(time, pp)
+    backend.openssl_assert(r > 0)
+    pp = backend._ffi.gc(
+        pp, lambda pointer: backend._lib.OPENSSL_free(pointer[0])
+    )
+    return pp, r
+
+
+def _encode_certificate_issuer(backend, certificate_issuer):
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Is this different from `_encode_alt_name`?

Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: It is not. Removing.
